contrib/NeonVM/cell.c

```diff
@@ -0,0 +1,384 @@
+        if (c->array == NULL) {
+            fatal_error("Unable to reallcoate memory for write array.");
+        }
+        if (c->array_size) {
+            for (n = c->array_size; n < i; n++) {
+                cell_resetCell(&c->array[n]);
+            }
+            c->array->array_size = i;
+        } else {
+            c->array_size++;
+        }
+    }
+    return &c->array[i];
+}
+
+Cell *cell_arrayForWrite(Cell *c)
+{
+    if (c->type == cNothing) {
+        c = cell_createArrayCell(1);
+    }
+    assert(c->type == cArray);
+    if (c->array == NULL) {
+        c->array = malloc(sizeof(Cell));
+        if (c->array == NULL) {
+            fatal_error("Unable to allocate memory for write array.");
+        }
+    }
+    return c->array;
+}
+
+Cell *cell_createDictionaryCell(uint64_t iEntries)
+{
+    Cell *c = cell_newCell();
+
+    c->type = cDictionary;
+    c->dictionary_size = iEntries;
+
+    // ToDo: Implement dictionary Cell Types
+    assert(c->type != cDictionary);
+
+    return c;
+}
+
+Cell *cell_createStringCell(uint64_t length)
+{
+    Cell *c = cell_newCell();
+
+    c->type = cString;
+    c->string = string_newString();
+    c->string->length = length;
+    c->string->data = malloc(length);
+    if (!c->string->data) {
+        fatal_error("Unable to allocate requested string with length 0x%08X.", length);
+    }
+    return c;
+}
+
+void cell_copyCell(Cell *dest, const Cell *source)
+{
+    assert(source != NULL);
+    assert(dest != NULL);
+
+    cell_clearCell(dest);
+
+    dest->number = source->number;
+    if (source->type == cString && source->string != NULL) {
+        dest->string = string_copyString(source->string);
+        //dest->string->length = source->string->length;
+        //dest->string = malloc(source->string->length);
+        //if (!dest->string) {
+        //    fatal_error("Failed to allocate memory for cell string copy.");
+        //}
+        //memcpy(dest->string, source->string, source->string->length);
+    } else {
+        dest->string = NULL;
+    }
+    dest->array_size = source->array_size;
+    if (source->type == cArray && source->array != NULL) {
+        uint32_t i = 0;
+        dest->array = malloc(sizeof(Cell) * dest->array_size);
+        for (i = 0; i < dest->array_size; i++) {
+            cell_copyCell(&dest->array[i], &source->array[i]);
+        }
+    } else {
+        dest->array = NULL;
+    }
+    dest->address = source->address;
+    dest->boolean = source->boolean;
+    dest->type = source->type;
+}
+
+int32_t cell_compareCell(const Cell * s, const Cell * d)
+{
+    assert(s->type == d->type);
+    // ToDo: Fix this, so that it actually COMPARES two cells, and returns 0 if equal, -1 if less than, and > 0 if greater than.
+    if (s->type != d->type) {
+        return -1;
+    }
+
+    switch (s->type) {
+        case cString:   return string_compareString(s->string, d->string);
+        case cAddress:  return s->address == d->address != 0;
```

Comment:
I don't understand the motivation behind the pattern "x == y != 0". What is that doing? Why not just "x == y"?